### ISSUE TITLE
Add configuration for waits/sleeps (linear or exponential) between retries

### ISSUE THREAD

**Issue by mjohnst** [elastic/elasticsearch-net#1499]
Using elasticsearch-net in our Elasticsearch Client. 

We are receiving intermittent MaxRetryExceptions that are not exposing any InnerExceptions (is set to null). We know that these retries are likely due to a network issue that is beyond our control anyway.

Instead of returning nothing for these attempted requests, we would rather return something after ~1 second (worst case).

Currently, elasticsearch-net retries up to our MaxRetries (10) in less than 200ms. We would like the retries to have a configurable back-off to allow for the network issue to correct before throwing MaxRetryExceptions. This way we have the option to wait between retries to fit our needs.

**Comment by Mpdreamz** [comment]
Hey @mjohnst thanks for opening this issue, will let it bake over the weekend and think about this. 

My first initial reaction is that this kind of logic might be best handled in the most opaque fashion around the client call in your application layer using something like [Poly](https://github.com/michael-wolfenden/Polly) which has support for this baked in. 

Retries are really meant to retry on different nodes in the cluster but still fail as fast as possible.

**Comment by mjohnst** [comment]
Hi @Mpdreamz thanks for the response! 

We are relatively confident that the error causing these retries is at the elasticsearch-net layer of the application since the requests are failing in a matter of milliseconds (we'd expect that a network level failure would take longer (x10 for each retry)).

Given this, we would prefer a option to set in elasticsearch-net for the back-off over using something like Poly that would encapsulate our requests, but I appreciate you guys thinking about it :)

**Comment by Mpdreamz** [comment]
Can you tell me a bit about your connection pool and how its set up? Are you using sniffing?

If you are not seeing any innerexceptions my first gut feeling is that the client is seeing 503 or 502 responses could this be the case?

**Comment by mjohnst** [comment]
I've added additional logging to the `RetryRequestAsyc` method to hopefully get some more information, but the only message we get is `TraceSource="w3wp.exe` Could this be a 502 or 503 surfacing?

I also added a increasing wait during retries (100ms * retry count), but the error persisted to our max retry count with these waits. This occurs intermittently for a few hours then not at all for up to a few days.

We are using sniffing for our connections.
Here is our code setting up the connection pool. Let me know if any other details would help.

```c#
private bool UpdateEndpoints(ElasticClient seedClient)
{
    try
    {
        var uris = this.GetHttpEndpoints(seedClient).ToList();

        if (uris.Count == 0)
        {
            Trace.TraceError("Unable to retrieve any IPs from seed endpoints");
            return false;
        }

        if (uris.All(uri => this.clientUris.Contains(uri)) && this.clientUris.All(clientUri => uris.Contains(clientUri)))
        {
            return true;
        }
        else
        {
            Trace.TraceInformation("Client endpoints updated");
            this.clientUris.Clear();
            foreach (var uri in uris) this.clientUris.Add(uri);
        }

        var pool = new StaticConnectionPool(uris);
        var settings = new ConnectionSettings(pool);
        if (this.enableDebugInfo)
        {
            settings.EnableMetrics(true);
            settings.ExposeRawResponse(true);
        }
        settings.MaximumRetries(this.numRetries);
        settings.SetDefaultPropertyNameInferrer(p => p);
        this.elasticClient = new ElasticClient(settings);

    }
    catch (Exception ex)
    {
        Trace.TraceError("Unable to update Elasticsearch endpoints: {0}; retrying", ex.ToString());
        return false;
    }
    finally
    {
        // Trigger next update
        if (this.updateEndpointsTimer != null)
            this.updateEndpointsTimer.Change(this.updateEndpointsInterval, Timeout.InfiniteTimeSpan);
    }
    return true;
}
```

**Comment by Mpdreamz** [comment]
HI @mjohnst,

I'm interested to see your GetHttpEndpoints() method, we also have a SniffingConnection that will sniff each node until one responds with the current cluster information.

http://nest.azurewebsites.net/elasticsearch-net/cluster-failover.html

Setting `SniffOnStartup` will fetch all the nodes from the cluster from the seed nodes you pass in the constructor. Setting `SniffOnConnectionFault` will resniff every time a failed request happens

You can also set `SniffLifeSpan` to resniff every `this.updateEndpointsInterval`. 

We are planning to make the sniff itself fully customizable in 2.0 but this is something we could totally backport in a 1.8 release.

**Comment by mjohnst** [comment]
Since our nodes have data, master, and http endpoints, we use this method to filter our node endpoints to just http.

Here is our `GetHttpEndpoints` method:
```c#
private IEnumerable<Uri> GetHttpEndpoints(ElasticClient seedClient)
{
    var nodesInfo = GetElasticsearchNodesInfo(seedClient);

    if (nodesInfo != null)
    {
        var uris =
            nodesInfo.Values
                .Where(
                    n =>
                        n.Settings.ContainsKey("node") &&
                        n.Settings["node"].Value["http"] == "true" &&
                        n.HttpAddress.StartsWith("inet[/") &&
                        n.HttpAddress.EndsWith("]"))
                .Select(n => "http://" + n.HttpAddress.Substring(6, n.HttpAddress.Length - 7));

        foreach (var uriString in uris)
        {
            Uri uri;
            if (Uri.TryCreate(uriString, UriKind.Absolute, out uri))
            {
                yield return uri;
            }
        }
    }
}
```
```c#
public virtual Dictionary<string, NodeInfo> GetElasticsearchNodesInfo(ElasticClient seedClient)
{
    if (seedClient != null)
        return seedClient.NodesInfo().Nodes;
    else
        return null;
}
```

**Comment by mjohnst** [comment]
Hey @Mpdreamz, thought I'd give an update to where we are and see if you had any suggestions to help us further expose what's going wrong in our client.

We tried to mitigate the MaxRetryExceptions by switching from sniffing to static (using the default Azure load balancer). Since then, we've stopped seeing the MaxRetryExceptions, but have been seeing 503 return codes in their place. You mentioned above that you suspected this could be the root cause. Do you have any thoughts on how we could further diagnose this or expose more information about what's causing the 503? (we've also contacted our Azure support to see if they can help)

Thanks!

**Comment by Mpdreamz** [comment]
Thanks for the update @mjohnst, much appreciated. 

Elasticsearch will throw a 503 if the node you are currently talking too can't find its master, It used to also throw a 503 on threadpool rejections (too many requests) but since mid 2014 ES switched 429 for that. 

https://github.com/elastic/elasticsearch/issues/6627

Another reason for the a 503 could be when all shards are in a transient recovery state and none can handle a search request (very rare). 

Or it could be the azure loadbalancer or another piece of infra. 

On `ConnectionSettings` there is a `SetConnectionStatusHandler` that you can pass a `Func` that will get called each time a good or bad response is being created. You could use that to specifically listen for 503's and log them so we can see what happens? Be sure to also set `KeepRawResponses` on Connectionsettings so that you can see it on `response.ConnectionStatus`

**Comment by mjohnst** [comment]
Adding some logging via the `SetConnectionStatusHandler` today. I'll let you know if we expose any more details.

**Comment by mjohnst** [comment]
We were able to get a bit more information from the logging.
We can now see this exception is part of the problem.

The `IElasticsearchResponse` `RawResponse`:
```json
{"error":"ClusterBlockException[blocked by: [SERVICE_UNAVAILABLE/1/state not recovered / initialized];]","status":503}
```
If you have any ideas about this, I'd be happy to hear :)

**Comment by Mpdreamz** [comment]
Hey @mjohnst are you able to locate the full elasticsearch exception from the elasticsearch log itself?

**Comment by mjohnst** [comment]
We had to add some more logging to narrow down what node it was happening on, but we did find the root cause to why the node was unavailable in its logs:
```
failed to send join request to master...
reason...RemoteTransportException [internal:discovery/zen/join]]...
nested: NotSerializableTransportException[org.elasticsearch.transport.ConnectTransportException]
```

Restarting the node's machine fixed this issue (we had trouble restarting just elasticsearch on the node).

This seems like an issue beyond Elasticsearch-Net, but it would be a great mitigation if our requests would retry on a different node instead of hitting a similarly "dead" node until max retries is hit.

Do you know if there's a mirror for this issue https://github.com/elastic/elasticsearch/issues/11709 in NEST/ES.Net ElasticsearchClients. If so, do you know whereabouts we could make changes in the code to force selection of a new node when there's a 503 or ClusterBlockException?

**Comment by Mpdreamz** [comment]
Hi @mjohnst sorry for the disconnect on this issue. 

NEST already retries on 503, we have quite extensive unit tests in place to see that we do. Here is one where we test a ping succeeding [but the subsequent call returning in a 503 so we should fallover](https://github.com/elastic/elasticsearch-net/blob/develop/src/Tests/Elasticsearch.Net.Tests.Unit/Failover/Revival/RevivalTests.cs#L146)

So I'm still wondering what sequence of events cause a 503, Can you share your connection settings?

**Comment by mjohnst** [comment]
I don't think there's any issue with NEST not retrying on 503s *at all*, but potentially not retrying on different nodes (like issue I referenced in my last post). We were definitely seeing the same node being hit 10 times (our max retry) in a row with 503s returning each time.

The 503 in our case looks like it was just surfaced from the node being unable to communicate with the master. We're really sure how this happened, but restarting the node has fixed it. 

Our connection settings are pretty much the defaults. We have just a few settings configured:

```C#
settings.SetDefaultPropertyNameInferrer(p => p);
settings.EnableMetrics(true);
settings.ExposeRawResponse(true);
settings.MaximumRetries(numRetries); # 10 from XML below
```

Some more settings in XML that we read: 
```xml
<Setting name="LatencyThreshold" value="5000" />
<Setting name="EsUpdateIntervalInSeconds" value="120" />
<Setting name="EsMaxRetries" value="10" />
```

**Comment by Mpdreamz** [comment]
Hi @mjohnst closing this issue for now. 

If you are still having issues please reopen a new ticket and we would love to help you get to the bottom of this. 

NEST 2.0 will have a lot of improvements with regards to testability and exposure of what goes on under the covers.
